File: shared/orchestrator.py
#!/usr/bin/env python3
"""
Antigravity Orchestrator
Parallel task execution with caching and semaphore-based concurrency control
"""

import asyncio
import hashlib
import time
import logging
from pathlib import Path
from typing import Callable, Any, Tuple, Dict


class AntigravityOrchestrator:
    """Orchestrator adapted for 2026_Antigravity workspace"""
    
    def __init__(
        self, 
        project_root: str = "/Users/jenkim/Downloads/2026_Antigravity", 
        n: int = 5, 
        cache_ttl: int = 3600, 
        cache_max: int = 1000
    ):
        """
        Initialize orchestrator
        
        Args:
            project_root: Root directory of the workspace
            n: Maximum concurrent tasks
            cache_ttl: Cache time-to-live in seconds
            cache_max: Maximum cache entries
        """
        self.root = Path(project_root)
        self.sem = asyncio.Semaphore(n)
        self.cache: Dict[str, Dict[str, Any]] = {}
        self.ttl = cache_ttl
        self.max = cache_max
        self.logs_dir = self.root / "logs"
        self.logs_dir.mkdir(exist_ok=True)
        self.logger = self._setup_logger()

    def _setup_logger(self) -> logging.Logger:
        """Setup logging configuration"""
        logger = logging.getLogger('AntigravityOrchestrator')
        logger.setLevel(logging.INFO)
        
        # File handler
        handler = logging.FileHandler(
            self.logs_dir / f"orchestrator_{time.strftime('%Y%m%d')}.log"
        )
        handler.setFormatter(
            logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
        )
        logger.addHandler(handler)
        
        # Console handler
        console = logging.StreamHandler()
        console.setLevel(logging.INFO)
        console.setFormatter(logging.Formatter('%(levelname)s - %(message)s'))
        logger.addHandler(console)
        
        return logger

    def _key(self, func: Callable, *args, **kwargs) -> str:
        """
        Generate cache key from function and arguments
        
        Args:
            func: Function to cache
            *args: Positional arguments
            **kwargs: Keyword arguments
            
        Returns:
            MD5 hash of function signature
        """
        s = f"{func.__name__}:{args}:{sorted(kwargs.items())}"
        return hashlib.md5(s.encode()).hexdigest()

    def _clean(self) -> None:
        """Remove expired cache entries and enforce size limit"""
        now = time.time()
        
        # Remove expired entries
        self.cache = {
            k: v for k, v in self.cache.items() 
            if now - v['time'] < self.ttl
        }
        
        # Enforce size limit
        if len(self.cache) > self.max:
            items = sorted(self.cache.items(), key=lambda x: x[1]['time'])
            keep = int(self.max * 0.8)
            self.cache = dict(items[-keep:])
            self.logger.info(f"Cache cleaned: {len(self.cache)}/{self.max} entries")

    async def run(self, tasks: list) -> list:
        """
        Execute tasks in parallel with caching
        
        Args:
            tasks: List of (func, args, kwargs) tuples
            
        Returns:
            List of results in same order as tasks
        """
        self.logger.info(f"Starting {len(tasks)} tasks with max {self.sem._value} concurrent")
        
        async def work(func: Callable, args: tuple, kwargs: dict) -> Any:
            k = self._key(func, *args, **kwargs)
            
            # Check cache
            if k in self.cache:
                self.logger.debug(f"Cache hit: {func.__name__}")
                return self.cache[k]['data']

            # Execute with semaphore
            async with self.sem:
                self.logger.debug(f"Executing: {func.__name__}")
                
                # Support both sync and async functions
                if asyncio.iscoroutinefunction(func):
                    r = await func(*args, **kwargs)
                else:
                    r = func(*args, **kwargs)

                # Cache result
                self.cache[k] = {'data': r, 'time': time.time()}
                self._clean()
                return r

        results = await asyncio.gather(*[work(f, a, k) for f, a, k in tasks])
        self.logger.info(f"Completed {len(tasks)} tasks")
        return results

    def cache_stats(self) -> dict:
        """Get cache statistics"""
        return {
            'entries': len(self.cache),
            'max_size': self.max,
            'ttl': self.ttl,
            'usage_percent': (len(self.cache) / self.max * 100) if self.max > 0 else 0
        }

    def clear_cache(self) -> None:
        """Clear all cache entries"""
        self.cache = {}
        self.logger.info("Cache cleared")


# Example usage
if __name__ == "__main__":
    async def example_task(n: int) -> str:
        """Example async task"""
        await asyncio.sleep(0.1)
        return f"Task {n} completed"
    
    async def main():
        orch = AntigravityOrchestrator()
        
        # Create tasks
        tasks = [(example_task, (i,), {}) for i in range(10)]
        
        # Run in parallel
        results = await orch.run(tasks)
        print(results)
        
        # Show stats
        print(f"Cache stats: {orch.cache_stats()}")
    
    asyncio.run(main())

File: projects/Fail_fis-korea-dashboard/src/utils/orchestrator.ts
/**
 * Antigravity Orchestrator
 * Parallel task execution with intelligent caching and concurrency control
 */

interface Task<T> {
      fn: () => Promise<T>;
      key: string;
}

interface CacheEntry<T> {
      data: T;
      timestamp: number;
}

export interface OrchestratorStats {
      cacheHits: number;
      cacheMisses: number;
      totalTasks: number;
      cacheSize: number;
      cacheUsagePercent: number;
}

export class AntigravityOrchestrator {
      private cache: Map<string, CacheEntry<any>> = new Map();
      private readonly maxConcurrent: number;
      private readonly cacheTTL: number;
      private readonly cacheMaxSize: number;
      private stats = {
            cacheHits: 0,
            cacheMisses: 0,
            totalTasks: 0
      };

      constructor(
            maxConcurrent = 5,
            cacheTTL = 3600000, // 1 hour in ms
            cacheMaxSize = 1000
      ) {
            this.maxConcurrent = maxConcurrent;
            this.cacheTTL = cacheTTL;
            this.cacheMaxSize = cacheMaxSize;
      }

      /**
   * Clean expired cache entries and enforce size limit
   */
      private clean(): void {
            const now = Date.now();

            // Remove expired entries
            const entries = Array.from(this.cache.entries());
            this.cache.clear();

            for (const [key, entry] of entries) {
                  if (now - entry.timestamp < this.cacheTTL) {
                        this.cache.set(key, entry);
                  }
            }

            // Enforce size limit
            if (this.cache.size > this.cacheMaxSize) {
                  const sortedEntries = Array.from(this.cache.entries())
                        .sort((a, b) => a[1].timestamp - b[1].timestamp);
                  const keep = Math.floor(this.cacheMaxSize * 0.8);
                  this.cache = new Map(sortedEntries.slice(-keep));
            }
      }

      /**
       * Execute tasks in parallel with caching and concurrency control
       */
      async run<T>(tasks: Task<T>[]): Promise<T[]> {
            this.clean();
            this.stats.totalTasks += tasks.length;

            const execute = async (task: Task<T>): Promise<T> => {
                  // Check cache
                  const cached = this.cache.get(task.key);
                  if (cached && Date.now() - cached.timestamp < this.cacheTTL) {
                        this.stats.cacheHits++;
                        return cached.data as T;
                  }

                  // Execute task
                  this.stats.cacheMisses++;
                  const result = await task.fn();

                  // Cache result
                  this.cache.set(task.key, {
                        data: result,
                        timestamp: Date.now()
                  });

                  return result;
            };

            // Execute with concurrency limit
            const results: T[] = [];
            for (let i = 0; i < tasks.length; i += this.maxConcurrent) {
                  const batch = tasks.slice(i, i + this.maxConcurrent);
                  const batchResults = await Promise.all(batch.map(execute));
                  results.push(...batchResults);
            }

            return results;
      }

      /**
       * Get cache statistics
       */
      getStats(): OrchestratorStats {
            return {
                  cacheHits: this.stats.cacheHits,
                  cacheMisses: this.stats.cacheMisses,
                  totalTasks: this.stats.totalTasks,
                  cacheSize: this.cache.size,
                  cacheUsagePercent: (this.cache.size / this.cacheMaxSize) * 100
            };
      }

      /**
       * Clear cache
       */
      clearCache(): void {
            this.cache.clear();
            this.stats = {
                  cacheHits: 0,
                  cacheMisses: 0,
                  totalTasks: 0
            };
      }

      /**
       * Get cached value
       */
      getCached<T>(key: string): T | null {
            const cached = this.cache.get(key);
            if (cached && Date.now() - cached.timestamp < this.cacheTTL) {
                  return cached.data as T;
            }
            return null;
      }

      /**
       * Set cached value
       */
      setCached<T>(key: string, value: T): void {
            this.cache.set(key, {
                  data: value,
                  timestamp: Date.now()
            });
            this.clean();
      }
}

export default AntigravityOrchestrator;
